deflate.go

```diff
@@ -0,0 +1,78 @@
+package imap
+
+import (
+	"compress/flate"
+	"io"
+	"net"
+)
+
+// A CompressUnsuppportedError is returned by Client.Compress when the provided
+// compression mechanism is not supported.
+type CompressUnsupportedError struct {
+	Mechanism string
+}
+
+func (err CompressUnsupportedError) Error() string {
+	return "COMPRESS mechanism " + err.Mechanism + " not supported"
+}
+
+// Compression algorithms for use with COMPRESS extension (RFC 4978).
+const (
+	// The DEFLATE algorithm, defined in RFC 1951.
+	CompressDeflate = "DEFLATE"
+)
+
+type deflateConn struct {
+	net.Conn
+
+	r io.ReadCloser
+	w *flate.Writer
+}
+
+func (c *deflateConn) Read(b []byte) (int, error) {
+	return c.r.Read(b)
+}
+
+func (c *deflateConn) Write(b []byte) (int, error) {
+	return c.w.Write(b)
+}
+
+func (c *deflateConn) Flush() error {
+	if f, ok := c.Conn.(flusher); ok {
+		if err := f.Flush(); err != nil {
+			return err
+		}
+	}
+
+	return c.w.Flush()
+}
+
+func (c *deflateConn) Close() error {
+	if err := c.r.Close(); err != nil {
+		return err
+	}
+
+	if err := c.w.Close(); err != nil {
+		return err
+	}
+
+	return c.Conn.Close()
+}
+
+// CreateDeflateConn wraps net.Conn with flate.Writer and flate.Reader.
+//
+// It is used by COMPRESS extension implementation and it not a part of the
+// library API. Do not use it.
+func CreateDeflateConn(c net.Conn, level int) (net.Conn, error) {
```

Comment:
Should we expose this to users? Or should we define this in an `internal` package?

Reply:
Not seeing a reason to keep it public.

Reply:
https://github.com/emersion/go-imap/pull/342/commits/2d8e2f30deecf1b156d11628a700434a88c84afd